text/0000-reseeding-rand.md

```diff
@@ -0,0 +1,682 @@
+`XorShiftRng`) will look like:
+
+```rust
+pub struct ChaChaRng { ... }
+
+impl ChaChaRng {
+    pub fn from_seed(seed: &[u32]) -> ChaChaRng { ... } // seed: [u32; 4] for XorshiftRng
+}
+
+impl Random<FullRange> for ChaChaRng {
+    ...
+}
+impl Reader for ChaChaRng {
+    ...
+}
+impl Rng for ChaChaRng {
+    ...
+}
+```
+
+These names are chosen to be somewhat vague: `ChaChaRng` is actually,
+currently, an implementation of the ChaCha20 algorithm, but the "20"
+value could be able to be customised backwards-compatibly in future,
+with default type parameters. (Similarly Xorshift is strictly
+Xorshift128, but has parameters that could be tweaked.) E.g. in
+hypothetical syntax
+
+```rust
+struct ChaChaRng<static N: u32 = 20> {
+    /* state */
+}
+let default: ChaChaRng = ...;
+let faster: ChaChaRng<8> = ...;
+```
+
+
+#### Alternatives
+
+As per the Alternatives section of the [free functions] discussion,
+these could be named `SecureRng`, `BalancedRng` and `WeakRng` (or have
+wrapper types of those names). However, the precedent is to not do
+this, both in other parts of the Rust library (e.g. `collections`) and
+in other random libraries, e.g. [C++11's `random`][cpp_random] and
+[D's `std.random`][d_random] both use the algorithms' names.
+
+Note, `XorShiftRng` is renamed to `XorshiftRng` since the "Xorshift"
+capitalisation is the correct one.
+
+### Thread RNG
+
+There will be one other RNG-implementer: the thread RNG. This RNG is
+stored in thread-local storage (TLS) to avoid having to reinitialise
+it each time, it will be defined something like
+
+```
+struct ThreadRngInner {
+    inner: ChaChaRng,
+    generated_entropy: usize,
+}
+pub struct ThreadRng {
+    inner: Rc<RefCell<ThreadRngInner>>
+}
+```
+
+so that the internal state is shared between all instances of the
+random number generator.
+
+The `generated_entropy` field will record the number of bytes output
+by the RNG, and the whole `ThreadRng` will be reseeded from the
+operating system once this passes a certain threshold.
+
+#### Alternatives
+
+Do not regularly reseed from the operating system: only the OS RNG is
+designed for use in cryptographic circumstances and so the overhead
+may not be useful. Also, reseeding like this restricts the ability for
+a program to have forcibly deterministic execution by seeding the
+`ThreadRng`. Changing this is backwards compatible.
+
+This may make more sense to exist in `std::thread` or
+`std::thread_local`.
+
+### Alternatives (for "Provided RNGs")
+
+Repeating `Rng` in ever name may not be necessary, a scheme like the
+following may be interesting:
+
+- `Os`, `RandomDevice`, `Secure`, `SecureRng` (or just `OsRng`)
+- `ChaCha`
+- `Xorshift`, `XorShift`
+- `std::rand::ThreadLocal` (a little confusing), `std::thread_local::Rng`
+
+We currently provide a `StdRng` type as an alias for another
+generator, this type disappears in this RFC, but could be reintroduced
+as a direct wrapper for `ChaChaRng`.
+
+## `Copy` and `Clone`
```

Comment:
The `Copy` trait is *supposed* to be implemented in every case where `Clone` is implemented via a shallow copy. It is used in places like `Cell` where a concept of plain-old data is required. This is a **clear** abuse of the traits and is both damaging in itself and setting a dangerous precedent where useful functionality is left out due to a minority subjective style choice. In fact, there's no justification given here for the design choice beyond empty rhetoric claiming it is somehow "particularly dangerous" to make use of the core language features that everyone will understand.